### PR TITLE
Make sampling decision up front when a span is created

### PR DESCRIPTION
**Pull request by dazuma**
Previously I thought we could defer sampling decisions to the point when we export the entire trace, but now I realize sampling decisions have to happen up front because a static yea/nay decision must be propagated to remote spans. So now, call the sampler when a span is created, and require that span-scoped samplers be passed in at span creation.

### CODE REVIEW

File: lib/opencensus/trace/span_context.rb [hunk 1]
```diff
@@ -123,12 +123,19 @@ def to_trace_context_header
...
+      #     defaults to the trace configuration's default sampler.
+      #
       # @return [SpanBuilder] A SpanBuilder object that you can use to
       #     set span attributes and create children.
       #
-      def start_span name, skip_frames: 0
+      def start_span name, skip_frames: 0, sampler: nil
         child_context = create_child
-        span = SpanBuilder.new child_context, skip_frames: skip_frames + 1
+        sampler ||= OpenCensus::Trace::Samplers::DEFAULT
```
Comment (chingor13) [inline lib/opencensus/trace/span_context.rb]: With the current default (probability at 10%), it's unlikely the user will see any in-depth trees. Should we switch the local default sampler to `AlwaysSample`?

Comment (dazuma) [reply lib/opencensus/trace/span_context.rb]: The spec for the Probability sampler says that if the parent has been sampled, then the child shall also be sampled. (https://github.com/census-instrumentation/opencensus-specs/blob/master/trace/Sampling.md). I went ahead and implemented that.

FWIW the default probability for the Java sampler is 10^-4...

Comment (chingor13) [reply lib/opencensus/trace/span_context.rb]: It seems like if a user does specify a local sampler we'd override the sampling decision even if it's the same type of sampler. I would expect that if no local sampler is provided, that we'd inherit the decision from our parent.

File: lib/opencensus/trace/samplers/probability.rb [hunk 2]
```diff
@@ -28,12 +28,17 @@ def initialize rate, rng: nil
...
-          value = if span_context
-                    (span_context.trace_id % 0x10000000000000000).to_f /
-                      0x10000000000000000
-                  else
-                    @rng.rand
-                  end
+          value =
+            if span_context
+              if span_context.sampled?
+                0.0
```
Comment (chingor13) [inline lib/opencensus/trace/samplers/probability.rb]: short circuit returning true here might be easier to read

File: lib/opencensus/trace/span_context.rb (hunk 1 above)
Comment (chingor13) [reply lib/opencensus/trace/span_context.rb]: We can clarify the sampling decision inheriting later.

File: lib/opencensus/trace/samplers/probability.rb (hunk 2 above)
Comment (dazuma) [reply lib/opencensus/trace/samplers/probability.rb]: done

**Pull request merged**